blockpage.lua

```diff
@@ -0,0 +1,156 @@
+--
+-- SPDX-FileCopyrightText: (c) 2024 Ring Zero Desenvolvimento de Software LTDA
+-- SPDX-License-Identifier: GPL-2.0-only
+--
+
+local linux  = require("linux")
+local socket = require("socket")
+local pack   = require("dome/pack")
+local config = require("dome/config")
+
+local PACKET = socket.af.PACKET
+local RAW    = socket.sock.RAW
+
+local ETH_HLEN  = 14
+local ETH_P_ALL = 0x0003
+
+local IPPROTO_TCP = 0x06
+
+local RST = 1 << 2
+local ACK = 1 << 4
+
+local packer   = pack.packer
+local unpacker = pack.unpacker
+
+local blockpage = {}
+
+-- https://datatracker.ietf.org/doc/html/rfc1141
+local function checksum(data, size)
+	local sum = 0
+	for i = 1, size, 2 do
+		sum = sum + string.unpack(">I2", data, i)
+	end
+	if size % 2 == 1 then
+		sum = sum + string.unpack("I1", data, size)
+	end
+	sum = sum & 0xffffffff
+	while sum >> 16 ~= 0 do
+		sum = (sum >> 16) + sum & 0xffff
+	end
+	return ~sum & 0xffff
+end
+
+local function ethhdr(packet)
+	local str, be16 = unpacker(packet)
+	return {
+		dst     = str(0, 6),
+		src     = str(6, 6),
+		ethtype = be16(12),
+	}
+end
+
+local function iphdr(packet)
+	local str, u8, be16, be32 = unpacker(packet, ETH_HLEN)
+	return {
+		version  = u8(0) >> 4,
+		ihl      = u8(0) & 0xf,
+		tos      = u8(1),
+		tot_len  = be16(2),
+		id       = be16(4),
+		frag_off = be16(6),
+		ttl      = u8(8),
+		protocol = u8(9),
+		check    = be16(10),
+		saddr    = be32(12),
+		daddr    = be32(16),
+	}
+end
+
+-- https://datatracker.ietf.org/doc/html/rfc9293
+local function tcphdr(packet)
+	local ihl = packet:getbyte(ETH_HLEN) & 0xf
+	local base = ETH_HLEN + ihl * 4
+	local _, u8, be16, be32 = unpacker(packet, base)
+	return {
+		source  = be16(0),
+		dest    = be16(2),
+		seq     = be32(4),
+		ack_seq = be32(8),
+		doff    = u8(12) >> 4,
+		flags   = u8(13),
+		window  = be16(14),
+		check   = be16(16),
+	}, base
+end
+
+local function pseudohdr(packet)
+	local ip, tcp = iphdr(packet), tcphdr(packet)
+	return string.pack(">I4I4BBH", ip.saddr, ip.daddr, 0, IPPROTO_TCP, tcp.doff * 4)
+end
+
+local function edit_eth(packet)
+	local setstr, setbe16 = packer(packet, 0)
+	local eth = ethhdr(packet)
+
+	setstr(0, eth.src)
+	setstr(6, eth.dst)
+	setbe16(12, eth.ethtype)
+end
+
+local function edit_ip(packet, tcp_len)
+	local _, setu8, setbe16, setbe32 = packer(packet, ETH_HLEN)
+	local ip = iphdr(packet, ETH_HLEN)
+	local size = ip.ihl * 4
+
+	setbe16(2, size + tcp_len) -- tot_len = 40
+	setbe16(4, 0)              -- id = 0
+	setbe16(6, 0)              -- frag_off = 0
+	setbe32(12, ip.daddr)      -- saddr = daddr
+	setbe32(16, ip.saddr)      -- daddr = saddr
+	setbe16(10, 0)             -- checksum = 0
+
+	local header = packet:getstring(ETH_HLEN, size)
```

Comment:
but we are explicitly turning it into a string, right? I think that's my question.. if we have advantage to that here and then use `string.unpack` or if it would be easier/simpler to compute checksum using luadata primitives..

Reply:
notice that below, in `edit_tcp`, we have to do a further manipulation, catenating a pseudo header to a data segment and computing the checksum from there. if we were to use luadata I'd have to create a `data.new`, then calling `data:setstring` anyway, so I found it more straightforward to create a `checksum` that does not depend on `luadata`, only on `lua`. 

so to sum up, I see two things: 1) having `checksum` to use `string` makes it more usable as it depends only on lua and 2) is kind of simplifies things a tiny bit, because I don't have to create a new `luadata`. 

that said, I'll go over everything once I'm done with the task at hand, I'm already refactoring around, so I will revisit this point again, probably later today. In the meantime, let me know your thoughts on the above points, please.